### DISCUSSION TITLE
any changes made to get_element_facets(element_id) ?

### DISCUSSION THREAD

**Discussion by roehligas** [cwapi3d/cwapi3dpython#203] [General] [answered]
Hi,

one of our scripts throws an error that wasn't there before.

```
Traceback (most recent call last):
  File "K:\02 KONSTRUKSJON\04 CADWORK\Cadwork_lib\userprofil_30\3d\API.x64\GenerateDoorBeams\GenerateDoorBeams.py", line 73, in <module>
    ess.main()
  File "//dc-1/Felles\02 KONSTRUKSJON/04 CADWORK/Cadwork_lib/USERPROFIL_30/3d/API.x64/GenerateDoorBeams\GenerateDoorBeams_lib.py", line 315, in main
    create_doorbeams([ae])
  File "//dc-1/Felles\02 KONSTRUKSJON/04 CADWORK/Cadwork_lib/USERPROFIL_30/3d/API.x64/GenerateDoorBeams\GenerateDoorBeams_lib.py", line 266, in create_doorbeams
    for i in range(len(facet)-1):
TypeError: object of type 'cadwork.vertex_list' has no len()
```

Have changes been made to this function? Before a python list containing a list of point3d objects was returned, now it is it's own class.

I need some documentation on this new class so that i can update our scripts

**Comment by Brunner246** [comment, marked as answer]
Hi,
The cadwork list objects have been reworked to the SP2024 version so that they are "identical" to the C++ interfaces (sorry for breaking the script, that was not our intention). 
The len function has not been implemented (but we can still implement this). Alternatively, a count() method is available. If you only want to "iterate" through the list, you can use the range based approach (`for item in items`)

see also https://docs.cadwork.com/projects/cwapi3dpython/en/latest/documentation/brep_data/

```python
ec.get_active_identifiable_element_ids()
[62748057]

gc.get_element_facets(62748057)
<cadwork.facet_list object at 0x00000298CB374A70>

facet_list = gc.get_element_facets(62748057)

facet_list.count()
10
for facet in facet_list:
    print(facet)

    
<cadwork.vertex_list object at 0x00000298CB40B7F0>
<cadwork.vertex_list object at 0x00000298CB366AB0>
<cadwork.vertex_list object at 0x00000298CB3B0BF0>
<cadwork.vertex_list object at 0x00000298CB40DF70>
```

**Reply by roehligas** [reply]
Thanks for the help, i figured it out how to update the code.

I found another interesting "bug", when you access elements of the vertex_list with an higher index than the length, it returns [0,0,0].

```
for i in range(facet.count()-1):
                        vert1 = facet[i]
                        vert2 = facet[i+1]
```

I found it by accidentally forgetting the -1 in the range length.
```
for i in range(facet.count()):
                        vert1 = facet[i]
                        vert2 = facet[i+1]
```

Here it returns [0,0,0] as last vertex.


I recommend to discuss this in your expert group if it has any benefits to return the global zero point rather than throw an exception, it seems rather unlogic from my end without knowing the bigger context. :)

**Reply by Brunner246** [reply]
Oh yes, that should definitely not be the case. An exception would have to be (will be) thrown.

We will fix this with the next cadwork update and add the len() function aswell.
